dango/dex/src/execute.rs

```diff
@@ -573,6 +583,51 @@ fn clear_orders_of_pair(
                 &order,
             )?;
         }
+
+        // Calculate the volume in USD for the filled order
+        let base_asset_price = querier.query_price(oracle, &base_denom, None)?;
+        let volume = base_asset_price.value_of_unit_amount(filled)?.into_int(); // TODO: Better to store as Decimal?
+
+        // Get the previous volume for the user's address
+        let previous_volume = VOLUMES
+            .prefix(&order.user)
+            .values(storage, None, None, IterationOrder::Descending)
+            .next()
+            .transpose()?
+            .unwrap_or(Uint128::ZERO);
+
+        // Record trading volume for the user's address
+        VOLUMES.save(
+            storage,
+            (&order.user, current_time),
+            &previous_volume.checked_add(volume)?,
+        )?;
+
+        // Get the username from the user address
+        let username = querier
+            .query_wasm_path(
+                account_factory,
+                &dango_account_factory::ACCOUNTS.path(order.user),
+            )?
+            .params
+            .owner();
+
+        if let Some(username) = username {
+            // Get the previous volume for the user's username
+            let previous_volume = VOLUMES_BY_USER
+                .prefix(&username)
+                .values(storage, None, None, IterationOrder::Descending)
+                .next()
+                .transpose()?
+                .unwrap_or(Uint128::ZERO);
+
+            // Record trading volume for the user's username
+            VOLUMES_BY_USER.save(
+                storage,
+                (&username, current_time),
+                &previous_volume.checked_add(volume)?,
+            )?;
+        }
```

Comment:
Can you extract these to two separate functions? So that it looks like this:

```rust
fn clear_orders_of_pair(/* ... */) -> anyhow::Result<()> {
    // ...

    record_volume(/* ... */)?;
    record_volume_by_username(/* ... *.)
}
```